### PR TITLE
Stabilize custom cache handlers and changing memory size.

### PR DESCRIPTION
**Pull request by leerob**
This PR stabilizes the previously introduced experimental config options for providing a custom cache handler (for both ISR as well as the Data Cache) and for disabling or configuring the in-memory cache size. The example usage would be as follows:

```js
// next.config.js
module.exports = {
  cacheHandler: require.resolve('./cache-handler.js'),
  cacheMaxMemorySize: 0 // disable default in-memory caching
}
```

This PR also updates the documentation to better reflect how to use the custom cache handler when self-hosting. Further information will be added in a following PR that also includes a full example of a custom cache handler that implements `revalidateTag` as well as passing in custom cache tags. The API reference docs have been updated here, as well as a version history added.

I also noticed that we currently have two duplicated versions of the ISR docs in the Pages Router docs: both for rendering and for data fetching. Data Fetching is the correct location for this page. There were no other references to the rendering version in the docs, so that must have been an accident. I'll need to a get a redirect going for that regardless.

Tests have been updated for `cacheHandler` and I added a new test for `cacheMaxMemorySize`.

### CODE REVIEW

**Comment by timneutkens** [comment]
Converted this to draft as the file has to be passed through the compiler before marked as stable.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 48ff7c6886364ee06346eb6ba9eb2a5fe26a4661

`pnpm test-start test/production/standalone-mode/required-server-files/required-server-files-app.test.ts` 
- required server files app router > should properly handle prerender for bot request
- required server files app router > should properly handle fallback for bot request
- required server files app router > should send cache tags in minimal mode for ISR
- required server files app router > should not send cache tags in minimal mode for SSR
- required server files app router > should not send invalid soft tags to cache handler

<details>
<summary>Expand output</summary>

  ● required server files app router › should properly handle prerender for bot request

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      114 |     })
      115 |
    > 116 |     expect(res.status).toBe(200)
          |                        ^
      117 |     const html = await res.text()
      118 |     const $ = cheerio.load(html)
      119 |

      at Object.toBe (production/standalone-mode/required-server-files/required-server-files-app.test.ts:116:24)

  ● required server files app router › should properly handle fallback for bot request

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      141 |     })
      142 |
    > 143 |     expect(res.status).toBe(200)
          |                        ^
      144 |     const html = await res.text()
      145 |     const $ = cheerio.load(html)
      146 |

      at Object.toBe (production/standalone-mode/required-server-files/required-server-files-app.test.ts:143:24)

  ● required server files app router › should send cache tags in minimal mode for ISR

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      182 |         redirect: 'manual',
      183 |       })
    > 184 |       expect(res.status).toBe(200)
          |                          ^
      185 |       expect(res.headers.get('x-next-cache-tags')).toBe(tags)
      186 |     }
      187 |   })

      at Object.toBe (production/standalone-mode/required-server-files/required-server-files-app.test.ts:184:26)

  ● required server files app router › should not send cache tags in minimal mode for SSR

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      197 |         redirect: 'manual',
      198 |       })
    > 199 |       expect(res.status).toBe(200)
          |                          ^
      200 |       expect(res.headers.get('x-next-cache-tags')).toBeFalsy()
      201 |     }
      202 |   })

      at Object.toBe (production/standalone-mode/required-server-files/required-server-files-app.test.ts:199:26)

  ● required server files app router › should not send invalid soft tags to cache handler

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      217 |         }
      218 |       )
    > 219 |       expect(res.status).toBe(200)
          |                          ^
      220 |       expect(res.headers.get('x-next-cache-tags')).toBeFalsy()
      221 |     }
      222 |   })

      at Object.toBe (production/standalone-mode/required-server-files/required-server-files-app.test.ts:219:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| buildDuration | 10.1s | 10.2s | N/A |
| buildDurationCached | 5.9s | 6.5s | ⚠️ +604ms |
| nodeModulesSize | 175 MB | 175 MB | N/A |
| nextStartRea..uration (ms) | 420ms | 421ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| 199-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | ✓ |
| 494.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 33.1 kB | 33.1 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.9 kB | 98.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| _app-HASH.js gzip | 205 B | 205 B | ✓ |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 505 B | 507 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.71 kB | 2.7 kB | N/A |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 319 B | 320 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 885 B | 885 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.06 kB | 1.06 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96.4 kB | 96.4 kB | N/A |
| page.js gzip | 141 kB | 141 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js cache-extravaganza | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 625 B | N/A |
| middleware-r..fest.js gzip | 148 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 48ff7c6886364ee06346eb6ba9eb2a5fe26a4661

**Review by vagnermaltauro — APPROVED** [review]

File: packages/next/src/server/config-schema.ts [hunk 1]
```diff
@@ -117,6 +117,8 @@ export const configSchema: zod.ZodType<NextConfig> = z.lazy(() =>
     analyticsId: z.string().optional(),
     assetPrefix: z.string().optional(),
     basePath: z.string().optional(),
+    cacheHandler: z.string().optional(),
```
Comment (A7med3bdulBaset) [inline packages/next/src/server/config-schema.ts:120]: `cacheHandler: ""` is a valid `ZodString` but it's neither of relative or absolute path. 

```suggestion
    cacheHandler: z.string().min(1).optional(),
```

**Review by vagnermaltauro — APPROVED** [review]

**Review by huozhi — APPROVED** [review]

**Pull request merged**